kong/cmd/utils/process.lua

```diff
@@ -0,0 +1,316 @@
+
+local tonumber = tonumber
+local type = type
+local floor = math.floor
+
+
+-- not supporting other usage of kill(2) for the moment
+local MIN_PID = 1
+
+-- source: proc(5) man page
+local MAX_PID = 2 ^ 22
+
+
+local E_EMPTY_PID_FILE   = "PID file is empty"
+local E_INVALID_PID_FILE = "PID file does not contain a valid PID"
+local E_NO_SUCH_PID_FILE = "PID file does not exist"
+local E_INVALID_PID_TYPE = "invalid PID type"
+local E_INVALID_PID      = "PID must be an integer from "
+                           .. MIN_PID .. " to " .. MAX_PID
+
+-- these are returned (and potentially passed through) from resty.signal
+local E_NO_SUCH_PROCESS = "No such process"         -- ESRCH
+local E_NOT_PERMITTED   = "Operation not permitted" -- EPERM
+
+
+local SIG_NONE   = 0
+local SIG_HUP    = 1
+local SIG_INT    = 2
+local SIG_QUIT   = 3
+local SIG_ILL    = 4
+local SIG_TRAP   = 5
+local SIG_ABRT   = 6
+local SIG_BUS    = 7
+local SIG_FPE    = 8
+local SIG_KILL   = 9
+local SIG_USR1   = 10
+local SIG_SEGV   = 11
+local SIG_USR2   = 12
+local SIG_PIPE   = 13
+local SIG_ALRM   = 14
+local SIG_TERM   = 15
+local SIG_CHLD   = 17
+local SIG_CONT   = 18
+local SIG_STOP   = 19
+local SIG_TSTP   = 20
+local SIG_TTIN   = 21
+local SIG_TTOU   = 22
+local SIG_URG    = 23
+local SIG_XCPU   = 24
+local SIG_XFSZ   = 25
+local SIG_VTALRM = 26
+local SIG_PROF   = 27
+local SIG_WINCH  = 28
+local SIG_IO     = 29
+local SIG_PWR    = 30
+local SIG_EMT    = 31
+local SIG_SYS    = 32
+local SIG_INFO   = 33
+
+
+---@param n any
+---@return integer
+local function validate_pid(n)
+  n = tonumber(n)
+  return n
+     -- good enough integer check for our use case
+     and floor(n) == n
+     and n >= MIN_PID and n <= MAX_PID
+     and n
+end
+
+
+---
+-- Read and return the process ID from a pid file.
+--
+---@param  fname       string
+---@return integer|nil pid
+---@return nil|string  error
+local function pid_from_file(fname)
+  local data, err = read_file(fname)
+  if not data then
+    -- map penlight error to our own
+    if err:lower():find("no such file", nil, true) then
+      err = E_NO_SUCH_PID_FILE
+    end
+
+    return nil, err
+  end
+
+  -- strip whitespace
+  data = data:gsub("^%s*(.-)%s*$", "%1")
+
+  if #data == 0 then
+    return nil, E_EMPTY_PID_FILE
+  end
+
+  local pid = validate_pid(data)
+
+  if not pid then
+    return nil, E_INVALID_PID_FILE
+  end
+
+  return pid
+end
+
+
+---
+-- Detects a PID from input and returns it as a number.
+--
+---@param  target      string|number
+---@return integer|nil pid
+---@return nil|string  error
+local function get_pid(target)
+  local typ = type(target)
+
+  local pid, err
+
+  if typ == "number" then
+    pid = target
+
+  elseif typ == "string" then
+    -- for the sake of compatibility we're going to accept PID input as a
+    -- numeric string, but this is potentially ambiguous with a PID file,
+    -- so we'll try to load from a file first before attempting to treat
+    -- the input as a numeric string
```

Comment:
Why not check the string's format instead?  Opening a file seems wasteful and requires checking for specific errors based on string values.